### DISCUSSION TITLE
Windows: Is it possible to bind to hostnames like *.localhost for local development?

### DISCUSSION THREAD

**Discussion by johanndev** [dotnet/yarp#826] [General] [answered]
Hi!

I'm wondering if YARP enables easier binding to hostnames during development of ASP.NET Core apps on Windows? For example, during development I would like my application to be reachable via https://dev.mysite.localhost. Currently, I'm aware of these options:

1. Via IIS 'proper': Install IIS locally, create website + self-signed cert and configure binding for my hostname: Setup is pretty easy, the DX is unfortunately not great, as debugging requires elevated privileges and changes must be deployed via `dotnet publish`, which forces me to temporarily stop the AppPool, as the dll gets locked by the worker process. Overall quite a lacking experience.

2. Via IIS Express: With changes to `applicationhost.config` and some `netsh` magic (see [here](https://stackoverflow.com/a/5186680/4173695)), I'm getting a pretty good DX including F5 debugging etc. Unfortunately, the setup is not reliable in my experience, as the whole setup sometimes stops working on some machines and working flawlessly on others; I'm of course not excluding the possibilty of the PEBKAC, at least in some cases ;)

3. Via YARP + IIS: I create a new proxy project according to the [Direct Proxying](https://microsoft.github.io/reverse-proxy/articles/direct-proxying.html) documentation, configure this project to proxy all my request to my `mysite` project running at e.g. https://localhost:5000, and enjoy easy F5 debugging <- This is my preferred solution at the moment because the setup is pretty easy and reliable.

What I really want: Solution 3., but without the need for IIS :) Can YARP help with this? I understand that at some point, admin privileges are needed, because otherwise the app won't be able to bind to port 443... some way to interface with http.sys would be needed.

Maybe there is a simpler solution that I'm not aware of at the moment, so I would greatly appreciate any pointers in the right direction - Thanks!

EDIT: Why do I want this in the first place? Mostly for enabling easier handling of OAUTH redirect URLs, especially in cases where I'm not able to configure them myself.

**Comment by Tratcher** [comment]
Note YARP plugs into the existing APS.NET Core routing system, it does not change anything about the existing server binding behaviors.

Have you considered using Kestrel instead of IIS? It registers by IP and port, not host names, and those can be configured programatically. Your main conflict would be port 443, but for development you should be able to use alternate ports?

**Comment by johanndev** [comment]
I regularly use Kestrel directly (or IIS Express), bound to port + IP. It is actually my preferred approach with most projects, as it involves no extra setup 😊

But it would be nice to be able to use custom hostnames during development in some cases. As I mentioned above, being able to use 'proper' hostnames for redirect URIs (e.g., when interfacing with an Azure B2C instance managed by a client) just feels nicer. It is in no way a deal breaker, as there are plenty of workarounds and solutions. 

I just want to make sure that I'm not overlooking some obvious and easy option.

**Reply by Tratcher** [reply]
The hostnames are mainly a DNS issue at that point. Have you looked at tools like `https://readme.localtest.me/`?

**Reply by johanndev** [reply]
Using this tool would restrict me to a single running app at a time (which would suffice in most cases), but it also wouldn't allow me to bind to port 443.

**Comment by samsp-msft** [comment]
You can cheat and add hostname entries directly to the etc/hosts file on both linux and windows. They will be resolved without DNS lookups.

**Reply by johanndev** [reply]
Editing the hosts file does not enable me to bind to https://dev.mysite.localhost on the standard 443 port - I would still need to use https://dev.mysite.localhost:4430 or some other port :)

**Reply by Tratcher** [reply]
Resolving host names and binding to ports are distinct concerns, until you get to http.sys/netsh/iis anyways. Binding to port 443 is always going to be your most limiting factor.

**Reply by johanndev** [reply]
I'm afraid that's true :/

**Comment by johanndev** [comment, marked as answer]
Thanks for all your input - I guess there is currently no 'ideal' solution for my scenario.
After thinking about the suggestions, I came to the realization that what I really want is a IIS-esque product that is tightly integrated with the ASP.NET Core ecosystem... Maybe a global tool that allows for commands àla: 
`dotnet hosting site add --name "MySite" --path "C:\inetpub\mysite"`
and
`dotnet hosting binding add --site "MySite" --hostname "https://dev.mysite.localhost" --type https --cert-thumbprint 1128a412342170694be96cb987c5c93af063d1c7`  
...or something along these lines. 

The IISAdministration cmdlets already enable the configuration of IIS through the CLI, so I guess in the end there is no *real* need for yet another product. But a uniform, cross-platform experience sure would be nice.

**Reply by Tratcher** [reply]
cc: @shirhatti

**Reply by shirhatti** [reply]
I'll be happy to review the IIS bits if you build something, but I'm definitely not building something 😄
